wp-lazy-loading.php

```diff
@@ -118,57 +120,115 @@ function wp_lazy_loading_enabled( $tag_name, $context ) {
+function wp_filter_content_attachment_images( $content, $context = null ) {
 	if ( null === $context ) {
 		$context = current_filter();
 	}
 
-	if ( ! wp_lazy_loading_enabled( 'img', $context ) ) {
+	if ( ! preg_match_all( '/<img [^>]+>/', $content, $matches ) ) {
 		return $content;
 	}
 
-	return preg_replace_callback(
-		'/<img\s[^>]+>/',
-		function( array $matches ) use( $content, $context ) {
-			if ( ! preg_match( '/\sloading\s*=/', $matches[0] ) ) {
-				$tag_html = $matches[0];
-
-				/**
-				 * Filters the `loading` attribute value. Default `lazy`.
-				 *
-				 * Returning `false` or an empty string will not add the attribute.
-				 * Returning `true` will add the default value.
-				 *
-				 * @since (TBD)
-				 *
-				 * @param string $default The filtered value, defaults to `lazy`.
-				 * @param string $tag_html The tag's HTML.
-				 * @param string $content The HTML containing the image tag.
-				 * @param string $context Optional. Additional context. Defaults to `current_filter()`.
-				 */
-				$value = apply_filters( 'wp_set_image_loading_attr', 'lazy', $tag_html, $content, $context );
+	$selected_images = array();
+	$attachment_ids  = array();
 
-				if ( $value ) {
-					if ( ! in_array( $value, array( 'lazy', 'eager' ), true ) ) {
-						$value = 'lazy';
-					}
+	foreach ( $matches[0] as $image ) {
+		if ( preg_match( '/wp-image-([0-9]+)/i', $image, $class_id ) ) {
+			$attachment_id = absint( $class_id[1] );
 
-					return str_replace( '<img', '<img loading="' . $value . '"', $tag_html );
-				}
+			if ( $attachment_id ) {
+				/*
+				 * If exactly the same image tag is used more than once, overwrite it.
+				 * All identical tags will be replaced later with 'str_replace()'.
+				 */
+				$selected_images[ $image ] = $attachment_id;
+
+				// Overwrite the ID when the same image is included more than once.
+				$attachment_ids[ $attachment_id ] = true;
 			}
+		}
+	}
+
+	if ( count( $attachment_ids ) > 1 ) {
+		/*
+		 * Warm the object cache with post and meta information for all found
+		 * images to avoid making individual database calls.
+		 */
+		_prime_post_caches( array_keys( $attachment_ids ), false, true );
+	}
+
+	$add_srcset_sizes = 'the_content' === $context;
+	$add_loading      = wp_lazy_loading_enabled( 'img', $context );
+
+	foreach ( $selected_images as $image => $attachment_id ) {
+		$image_meta = wp_get_attachment_metadata( $attachment_id );
+
+		$filtered_image = $image;
+
+		// Add 'srcset' and 'sizes' attributes if applicable.
+		if ( $add_srcset_sizes && false === strpos( $filtered_image, ' srcset=' ) ) {
+			$filtered_image = wp_image_add_srcset_and_sizes( $filtered_image, $image_meta, $attachment_id );
+		}
+
+		// Add 'loading' attribute if applicable.
+		if ( $add_loading && false === strpos( $filtered_image, ' loading=' ) ) {
+			$filtered_image = wp_image_add_loading( $filtered_image, $image_meta, $attachment_id, $content, $context );
+		}
```

Comment:
I agree with that sentiment, but there's also no harm in being conservative at the start and only covering attachment images. We can easily make this work also with other images too (pass attachment ID 0), but that doesn't mean we need to right away.

I don't have a strong opinion on this, but 3P images, e.g. tracking pixels, may have different implications. We assume there's no issue, but we don't know for sure. Covering 3P images here could easily become a follow-up (separate PR or even separate core release if needed).

Reply:
> no harm in being conservative

Right, but this is mostly an "artificial" limitation: "You get lazy loading of images but only if they are uploaded to WP and are inserted properly by using the editor. Other images will not be lazy-loaded because...." (we said so?) :)

But yes, I agree this is not a huge issue as the majority of images in post_content will have the attachment ID class (as it has been used for srcset and sizes for a while). Skipping images that don't have it will skip all "images from URLs" perhaps including images from CDNs, which would be pretty disappointing for the feature.